app/scripts/background.js

```diff
@@ -184,10 +185,38 @@ let connectExternal;
 browser.runtime.onConnect.addListener(async (...args) => {
   // Queue up connection attempts here, waiting until after initialization
   await isInitialized;
+  const remotePort = args[0];
+  const { sender, name } = remotePort;
 
-  // This is set in `setupController`, which is called as part of initialization
-  connectRemote(...args);
+  const url = sender?.url;
+  const detectedProcessName = url ? getEnvironmentType(url) : '';
+  const processName = name;
+
+  if (
+    processName === 'metamask-contentscript' &&
+    detectedProcessName === 'background'
+  ) {
+    // This is set in `setupController`, which is called as part of initialization
+    connectRemote(...args);
+    return;
+  }
+
+  if (detectedProcessName !== processName) {
+    throw new Error('UNEXPECTED_CONNECTION_ATTEMPT');
+  }
+
+  const connectionId = generateConnectionId(remotePort, processName);
+  const connection = openMetamaskConnections.get(connectionId);
+
+  if (Boolean(connection) === false) {
+    // This is set in `setupController`, which is called as part of initialization
+    connectRemote(...args);
+    openMetamaskConnections.set(connectionId, true);
+  } else {
+    throw new Error('CONNECTION_ALREADY_EXISTS');
```

Comment:
It seems like this would prevent multiple fullscreen windows from being open. But that's a valid use case that we want to continue supporting.